cloudinit/sources/DataSourceCloudCIX.py

```diff
@@ -0,0 +1,206 @@
+            setattr(self, "_net_cfg", None)
+
+    def _get_data(self):
+        """
+        Fetch the user data and the metadata
+        """
+        try:
+            crawled_data = util.log_time(
+                logfunc=LOG.debug,
+                msg="Crawl of metadata service",
+                func=self.crawl_metadata_service,
+            )
+        except sources.InvalidMetaDataException as error:
+            LOG.debug(
+                "Failed to read data from CloudCIX datasource: %s", error
+            )
+            return False
+
+        self.metadata = crawled_data["meta-data"]
+        self.userdata_raw = util.decode_binary(crawled_data["user-data"])
+
+        return True
+
+    def crawl_metadata_service(self) -> dict:
+        md_url = self.determine_md_url()
+        if md_url is None:
+            raise sources.InvalidMetaDataException(
+                "Could not determine metadata URL"
+            )
+
+        data = read_metadata(md_url, self.get_url_params())
+        return data
+
+    def determine_md_url(self) -> Optional[str]:
+        if self._metadata_url:
+            return self._metadata_url
+
+        # Try to reach the metadata server
+        url_params = self.get_url_params()
+        base_url, _ = url_helper.wait_for_url(
+            METADATA_URLS,
+            max_wait=url_params.max_wait_seconds,
+            timeout=url_params.timeout_seconds,
+        )
+        if not base_url:
+            return None
+
+        # Find the highest supported metadata version
+        for version in range(METADATA_VERSION, 0, -1):
+            url = url_helper.combine_url(
+                base_url, "v{0}".format(version), "metadata"
+            )
+            try:
+                response = url_helper.readurl(url, timeout=self.url_timeout)
+            except url_helper.UrlError as e:
+                LOG.debug("URL %s raised exception %s", url, e)
+                continue
+
+            if response.ok():
+                self._metadata_url = url_helper.combine_url(
+                    base_url, "v{0}".format(version)
+                )
+                break
+            else:
+                LOG.debug("No metadata found at URL %s", url)
+
+        return self._metadata_url
+
+    @staticmethod
+    def ds_detect():
+        return is_platform_viable()
+
+    @property
+    def network_config(self):
+        if self._net_cfg:
+            return self._net_cfg
+
+        if not self.metadata:
+            return None
+        self._net_cfg = self._generate_net_cfg(self.metadata)
+        return self._net_cfg
+
+    def _generate_net_cfg(self, metadata):
+        netcfg: Dict[str, Any] = {"version": 2, "ethernets": {}}
+        macs_to_nics = net.get_interfaces_by_mac()
+        nameservers = metadata["network"].get("nameservers", None)
+
+        for iface in metadata["network"]["interfaces"]:
+            name = macs_to_nics.get(iface["mac_address"])
+            routes = iface.get("routes", None)
```

Comment:
@jgrassler it appears you may have control over your datasource producing structured data for `network` config. I find it a bit strange that your IMDS example data surfaces "routes" content that is structured to as viable [netplan v2](https://netplan.readthedocs.io/en/stable/netplan-yaml/) values for routes and nameservers and you are essentially reconstructing that network v2 in cloud-init. If your datasource has the ability to provide full network: version: 2 config under the 'network' key, then it would give you flexibility to grow/change your network config in the future in CloudCIX server Instance Metadata Service directly without waiting on any changes (and published releases) from cloud-init. If you have control over that element in CloudCIX IMDS, that may be a slightly more flexible path for you in the future. Note if cloud-init sees `network: version: 2` from any IMDS and it is on a system with netplan, cloud-init passes that network config directly to netplan for writing and managing network configuration.

Reply:
@jgrassler This is not a  blocking review commit preventing us from landing this PR upstream in cloud-init, but it does make future uploads of CloudCIX a bit more complex if we have to worry about CloudCIX network config formats changing in the future and old distribution images in cloud-init not being able to consume a changed network config IMDS endpoint in CloudCIX.

Reply:
> @jgrassler it appears you may have control over your datasource producing structured data for `network` config.

Yes, we do indeed.

> I find it a bit strange that your IMDS example data surfaces "routes" content that is structured as viable [netplan v2](https://netplan.readthedocs.io/en/stable/netplan-yaml/) values for routes and nameservers and you are essentially reconstructing that network v2 in cloud-init.

That is correct, too. We structured it in that manner to reduce the translation work needed in `cloud-init`.

> If your datasource has the ability to provide full network: version: 2 config dictionary under the 'network' key, then it would give you flexibility to grow/change your network config in the future in CloudCIX server Instance Metadata Service directly without waiting on any changes (and published releases) from cloud-init. If you have control over that element in CloudCIX IMDS, that may be a slightly more flexible path for you in the future

We had a discussion about this today and we decided to go with it because we're not 100% sure what else we may or may not need to configure in the future either. I'm making and testing the change right now and will update the pull request later.